cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/condition/BasicConditionContext.java

```diff
@@ -18,27 +18,95 @@
+import com.google.common.collect.ImmutableMap;
 import org.apache.tephra.TransactionFailureException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.HashMap;
+import java.util.List;
 import java.util.Map;
 
 /**
  * Default implementation of the {@link ConditionContext}.
  */
 public class BasicConditionContext extends AbstractStageContext implements ConditionContext {
-
+  private static final Logger LOG = LoggerFactory.getLogger(BasicConditionContext.class);
   private final WorkflowContext context;
+  private final Map<String, StageStatistics> stageStatistics;
 
   public BasicConditionContext(WorkflowContext context, PipelineRuntime pipelineRuntime, StageSpec stageSpec) {
     super(pipelineRuntime, stageSpec);
     this.context = context;
+    this.stageStatistics = createStageStatistics(context);
+  }
+
+
+  private Map<String, StageStatistics> createStageStatistics(WorkflowContext context) {
+    Map<String, StageStatistics> stageStatistics = new HashMap<>();
+    WorkflowToken token = context.getToken();
+    for (WorkflowToken.Scope scope : Arrays.asList(WorkflowToken.Scope.SYSTEM, WorkflowToken.Scope.USER)) {
+      Map<String, List<NodeValue>> all = token.getAll(scope);
+      for (Map.Entry<String, List<NodeValue>> entry : all.entrySet()) {
+        if (!entry.getKey().startsWith(Constants.StageStatistics.PREFIX + ".")) {
+          continue;
+        }
+        String stageKey = entry.getKey().substring(Constants.StageStatistics.PREFIX.length() + 1);
+
+        String stageName;
+        if (stageKey.endsWith(Constants.StageStatistics.INPUT_RECORDS)) {
+          stageName = stageKey.substring(0, stageKey.length() - Constants.StageStatistics.INPUT_RECORDS.length() - 1);
+        } else if (stageKey.endsWith(Constants.StageStatistics.OUTPUT_RECORDS)) {
+          stageName = stageKey.substring(0, stageKey.length() - Constants.StageStatistics.OUTPUT_RECORDS.length() - 1);
+        } else if (stageKey.endsWith(Constants.StageStatistics.ERROR_RECORDS)) {
+          stageName = stageKey.substring(0, stageKey.length() - Constants.StageStatistics.ERROR_RECORDS.length() - 1);
+        } else {
+          // should not happen
+          LOG.warn(String.format("Ignoring key '%s' in the Workflow token while generating stage statistics " +
+                                   "because it is not in the form " +
+                                   "'stage.statistics.<stage_name>.<input|output|error>.records'.",
+                                 stageKey));
+          continue;
+        }
+
+        // Since stage names are unique and properties for each stage tracked are unique(input, output, and error)
+        // there should only be single node who added this particular key in the Workflow
+        long value = entry.getValue().get(0).getValue().getAsLong();
+
+
+        StageStatistics statistics = stageStatistics.get(stageName);
```

Comment:
doesn't look like you need to get this or do the null check below, since you're not adding counts to any existing object. Unless you're supposed to be adding below? But I don't think you can see the same key more than once right?